File: easy/max_product_of_3_numbers/max_product_of_3_numbers.go
package main

import (
	"fmt"
	"math"
)

func maximumProduct(nums []int) int {
	firstMin, secondMin := math.MaxInt32, math.MaxInt32
	firstMax, secondMax, thirdMax := math.MinInt32, math.MaxInt32, math.MinInt32

	for _, n := range nums {
		if n <= firstMin {
			firstMin, secondMin = n, firstMin
		} else if n <= secondMin {
			secondMin = n
		}

		if n >= firstMax {
			firstMax, secondMax, thirdMax = n, firstMax, secondMax
		} else if n >= secondMax {
			secondMax, thirdMax = n, secondMax
		} else if n >= thirdMax {
			thirdMax = n
		}
	}

	product1 := firstMin * secondMin * firstMax
	product2 := firstMax * secondMax * thirdMax

	if product1 > product2 {
		return product1
	}
	return product2
}

func main() {
	fmt.Println("Solution 1 : ", maximumProduct([]int{1, 2, 3}))
	fmt.Println("Solution 2 : ", maximumProduct([]int{1, 2, 3, 4}))
	fmt.Println("Solution 3 : ", maximumProduct([]int{-4, -3, -2, -1, 60}))
}

File: easy/max_product_of_3_numbers/max_product_of_3_numbers.py
class Solution(object):
    def maximumProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        first_min, second_min = float("inf"), float("inf")
        first_max, second_max, third_max = - \
            float("inf"), -float("inf"), -float("inf")

        for n in nums:
            if n <= first_min:
                second_min, first_min = first_min, n
            elif n <= second_min:
                second_min = n

            if n >= first_max:
                third_max, second_max, first_max = second_max, first_max, n
            elif n >= second_max:
                third_max, second_max = second_max, n
            elif n >= third_max:
                third_max = n

        product_1 = first_min*second_min*first_max
        product_2 = first_max*second_max*third_max
        return max(product_1, product_2)


s = Solution()
print("Solution 1 : ", s.maximumProduct([1, 2, 3]))
print("Solution 2 : ", s.maximumProduct([1, 2, 3, 4]))
print("Solution 3 : ", s.maximumProduct([-4, -3, -2, -1, 60]))
